receiver/logrus.go

```diff
@@ -0,0 +1,166 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+package receiver
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/telenornms/skogul"
+)
+
+// LogrusLog configures the logrus log receiver
+type LogrusLog struct {
+	Loglevel string
+	Handler  skogul.HandlerRef
+}
+
+var logrusLogLogger = logrus.New()
+
+// Proxy all log entries onto this logger so we can customize its output without polluting the global logger
+var proxyLogger = logrus.New()
+
+var logMetadataFields *[]string
+
+func (lg *LogrusLog) configureLogger() error {
+	// Set up internal logger so we don't cause recursive logging in case we log errors
+	logrusLogLogger.SetOutput(io.Writer(os.Stdout))
+	logrusLogLogger.WithFields(logrus.Fields{
+		"category": "receiver",
+		"receiver": "logrus",
+	})
+	loglevel := skogul.GetLogLevelFromString(lg.Loglevel)
+	logrusLogLogger.SetLevel(loglevel)
+
+	metadataFields := []string{"category", "level"}
+	a := make([]string, 0)
+	logMetadataFields = &a
+	for _, field := range metadataFields {
+		*logMetadataFields = append(*logMetadataFields, field)
+	}
+	return nil
+}
+
+// Parse logrus logs to a skogul.Container
+func (lg *LogrusLog) Parse(bytes []byte) (*skogul.Container, error) {
+	logrusLogLogger.Debug("Parsing log line")
+	var c *skogul.Container
+	err := json.Unmarshal(bytes, &c)
+	if err != nil {
+		logrusLogLogger.Error("Failed to marshal logrus log")
+	}
+	return c, nil
+}
+
+// Write logrus logs as skogul.Containers to a handler
+func (lg *LogrusLog) Write(bytes []byte) (int, error) {
+	metadata := make(map[string]interface{})
+
+	var data map[string]interface{}
+	if err := json.Unmarshal(bytes, &data); err != nil {
+		logrusLogLogger.Error("Failed to unmarshal logrus log for sending it to log receiver")
+		return 0, err
+	}
+
+	// Extract metadata fields from data
+	for _, field := range *logMetadataFields {
+		metadata[field] = data[field]
+	}
+
+	timestampData := data["time"].(string)
+	timestamp, err := time.Parse(time.RFC3339, timestampData)
+	if err != nil {
+		// @ToDo: Do we want to error out if this fails?
+		logrusLogLogger.Errorf("Failed to parse timestamp '%s' '%s'\n", "time", timestampData)
+	}
+
+	m := skogul.Metric{
+		Time:     &timestamp,
+		Metadata: metadata,
+		Data:     data,
+	}
+	c := skogul.Container{
+		Metrics: []*skogul.Metric{&m},
+	}
+
+	logEntry, err := json.Marshal(c)
+	if err != nil {
+		logrusLogLogger.Error("Failed to marshal logrus log into container")
+	}
+
+	lg.Handler.H.Handle(logEntry)
```

Comment:
In this case, there's not really any need to use json marshal/unmarshal any more. You already have the container, and TransformAndSend() exists for this reason. It was done in the original log-receiver because there was no structured data.

So you can remove the entire Parse()-function and skil line 111 through 115 and instead do lg.Handler.H.TransformAndSend(&c). Unless I'm missing something?

Reply:
That makes total sense 👍